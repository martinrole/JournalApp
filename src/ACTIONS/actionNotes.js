import { addDoc, collection, deleteDoc, doc, updateDoc } from "@firebase/firestore"
import Swal from "sweetalert2"
import { baseDatos } from "../FIREBASE/firebaseConfig"
import { fileUploadHelper } from "../HELPERS/fileUpload"
import { loadNotes } from "../HELPERS/loadNotes"
import { types } from "../TYPES/types"


export const newNoteAction = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth

        const newNote = {
            titulo: '',
            texto: '',
            fecha: new Date().getTime()
        }

        const docRef = await addDoc( collection(baseDatos, `${uid}/journal/notes`), newNote )
        //console.log('docRef: ', docRef);
        
        dispatch( activeNote(docRef.id, newNote) )
        dispatch( addNewNote(docRef.id, newNote) )
    }
}

export const loadNotesAction = ( uid ) => {
    return async ( dispatch ) => {
        const notes = await loadNotes( uid )
        dispatch( setNotes( notes ) )
    }
}

export const saveNoteAction = ( note ) => async ( dispatch, getState ) => {
        
    const { uid } = getState().auth

    //En los 3 puntos queda el resto de propiedades de la nota, y así se puede extraer el id para no repetirlo
    const {id, imgUrl = null, ...notaFinal } = note

    //Asi se envia a Firebase en la versiòn beta 9
    const noteRef = doc(baseDatos, `${uid}/journal/notes/${note.id}`)
    await updateDoc( noteRef, { imgUrl, ...notaFinal })

    dispatch( refreshNote( note.id, note ))
    Swal.fire('Saved', note.titulo, 'success' )
}

export const deleteNoteAction = ( id ) => async ( dispatch, getState ) => {

    const { uid } = getState().auth

    const  noteRef = doc(baseDatos, `${uid}/journal/notes/${id}`)
    await deleteDoc(noteRef)

    dispatch( deleteNote( id ) )
}

export const fileUpload = ( archivo ) => async ( dispatch, getState ) => {
    
    const activeNote = getState().notes.activeNote

    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait...',
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
            Swal.showLoading()
        }
    })

    
    const fileUrl = await fileUploadHelper( archivo )
    activeNote.imgUrl = fileUrl
    
    dispatch( saveNoteAction( activeNote ) )
    Swal.close()

}

export const addNewNote = ( id, nota ) => ({
    type: types.notesAdd,
    payload: { id, ...nota }
})

export const refreshNote = ( id, nota ) => ({
    type: types.notesUpdate,
    payload: { id, nota: { id, ...nota } }
})


export const activeNote = (id, nota) => ({
    type: types.notesActive,
    payload: { id, ...nota }
})

export const setNotes = ( nota ) => ({
    type: types.notesLoad,
    payload: nota
})

export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
})

export const clearNotes = () => ({
    type: types.notesClear
})
