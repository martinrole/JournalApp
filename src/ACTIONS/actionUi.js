import { types } from "../TYPES/types"

export const actionSetError = ( error ) => ({
    type: types.uiSetError,
    payload: error
})

export const actionRemoveError = () => ({
    type: types.uiRemoveError
})


export const actionLoading = (type) => ({
    type: type
})

