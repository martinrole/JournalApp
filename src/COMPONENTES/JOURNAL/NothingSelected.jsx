import React from 'react'

export const NothingSelected = () => {
    return (
        <div className="journal__nothing-content">
            <p>
                Select something
                <br />
                or create an entry!
            </p>

            <i className="far fa-star fa-4x mt-5"></i>
        </div>
    )
}
