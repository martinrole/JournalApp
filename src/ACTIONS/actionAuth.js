import { autorizacion, googleAuth } from "../FIREBASE/firebaseConfig";
import { signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { types } from "../TYPES/types"
import { actionLoading, actionSetError } from "./actionUi";
import Swal from 'sweetalert2'
import { clearNotes } from "./actionNotes";

//Middleware:
export const actionLoginAsync = ( correo, contrasena ) => {
    //Esta funciÃ³n va a retornar un Callback:
    //el dispatch lo va  a entregar el thunk instalado
    return (dispatch) => {

        dispatch( actionLoading(types.uiLoadingOn) )

        return signInWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( ({user}) => {
                dispatch( actionLogin( user.uid, user.displayName ) )
                dispatch( actionLoading(types.uiLoadingOff) )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
                dispatch( actionLoading(types.uiLoadingOff) )
            }) 
            
        }
}



export const actionRegisterAsync = ( nombre, correo, contrasena ) => {
    return ( dispatch ) => {

        return createUserWithEmailAndPassword( autorizacion, correo, contrasena )
            .then( async ( { user } ) => {
                await updateProfile( user, { displayName: nombre } )
                
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )
            })
            .catch ( error => {
                dispatch( actionSetError(error.code) )
                Swal.fire('Error', error.code, 'error')
            })  
    }
}

export const actionGoogleLogin = () => {
    return ( dispatch ) => {
        signInWithPopup( autorizacion, googleAuth )
            .then( ( { user } ) => {
                dispatch(
                    actionLogin( user.uid, user.displayName)
                )

            })     
    }
}

export const actionLogOutAsync = () => {
    return async ( dispatch ) => {
        await signOut(autorizacion)

        dispatch( actionLogOut() )
        dispatch( clearNotes() )
    }
}

//Una forma de hacerlo:
// export const actionAuth = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid,
//             displayName
//         }
//     }
// }


//Segunda forma de hacerlo:
export const actionLogin = (uid, displayName) => ({  
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const actionLogOut = () => ({
    type: types.logout
})