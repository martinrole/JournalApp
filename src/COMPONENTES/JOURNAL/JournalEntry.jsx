import React from 'react'
import { useDispatch } from 'react-redux'
import moment from 'moment'
import { activeNote } from '../../ACTIONS/actionNotes'

export const JournalEntry = ( { id, titulo, texto, fecha, imgUrl } ) => {

    const noteDate = moment(fecha)
    const dispatch = useDispatch()
    
    const handleNote = () => {
        dispatch( activeNote( id, { titulo, texto, fecha, imgUrl } ) )
    }

    return (
        <div className="journal__entry pointer" onClick={ handleNote }>

            <div className="journal__entry-body">
                <p className="journal__entry-title">{ titulo }</p>
                <p className="journal__entry-content">{ texto }</p>
            </div>

            <div className="journal__entry-date-box">
                <span>{ noteDate.format('dddd')}</span>
                <h4>{ noteDate.format('Do') }</h4>
            </div>
        </div>
    )
}
