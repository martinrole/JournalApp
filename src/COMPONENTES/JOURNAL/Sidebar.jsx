import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { actionLogOutAsync } from '../../ACTIONS/actionAuth'
import { newNoteAction } from '../../ACTIONS/actionNotes'
import { JournalEntries } from './JournalEntries'


export const Sidebar = () => {

    const { name } = useSelector(state => state.auth) 
    const dispatch = useDispatch()

    const handleLogOut = () => {
        dispatch( actionLogOutAsync() )
    }

    const handleAddNote = () => {
        dispatch( newNoteAction() )
    }

    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    <span> { name }</span>
                </h3>

                <button className="btn" onClick={ handleLogOut }>Log out</button>
            </div>

            <div className="journal__new-entry" onClick={ handleAddNote }>
                <i className="far fa-calendar-plus fa-3x"></i>
                <p className="mt-5">New Note</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
