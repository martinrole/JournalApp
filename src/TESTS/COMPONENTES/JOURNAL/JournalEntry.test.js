import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom'
import { mount } from 'enzyme';
import { Provider } from "react-redux";
import { JournalEntry } from '../../../COMPONENTES/JOURNAL/JournalEntry';
import { activeNote } from '../../../ACTIONS/actionNotes';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {}

const nota = {
    id: 123,
    titulo: 'Titulo de prueba',
    texto: 'El texto de prueba de la nota',
    fecha: 0,
    imgUrl: 'https://www.estaurlnoexiste,esdeprueba.com/foto.jpg'
}

let store = mockStore( initState )
store.dispatch = jest.fn()

const wrapper = mount( 
    <Provider store={ store }>
        <JournalEntry { ...nota }/>
    </Provider>
)

describe('Prueba en componente JournalEntry', () => {

    test('Debe de mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot()
    })

    test('Debe de activar la nota seleccionada', () => {
        wrapper.find('.journal__entry').prop('onClick')()
        expect( store.dispatch ).toHaveBeenCalledTimes(1)
        expect( store.dispatch ).toHaveBeenCalledWith(
            activeNote( nota.id, { ...nota } )
        )
    })
})