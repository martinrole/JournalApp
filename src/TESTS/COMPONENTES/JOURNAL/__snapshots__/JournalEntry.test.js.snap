// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en componente JournalEntry Debe de mostrar el componente correctamente 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Titulo de prueba
    </p>
    <p
      className="journal__entry-content"
    >
      El texto de prueba de la nota
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
